core/src/ldml/ldml_processor.cpp

```diff
@@ -59,23 +59,130 @@ namespace {
+
+  // validate size and offset
+  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize >  LDML_LENGTH_SECT);
+  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize <  (data.size() - sizeof(kmx::COMP_KEYBOARD_EX)));
+  KMXPLUS_ASSERT(true, ex->kmxplus.dpKMXPlus     >= sizeof(kmx::COMP_KEYBOARD_EX));
+  KMXPLUS_ASSERT(true, ex->kmxplus.dpKMXPlus     <  data.size());
+
+  // calculate pointers to start and end of kmxplus
+  const uint8_t * const kmxplusdata   = data.data() + ex->kmxplus.dpKMXPlus;      // Start of + data
+  // const uint8_t* kmxpluslimit  = kmxplusdata    + ex->kmxplus.dwKMXPlusSize;  // End of   + data
+
+  // Now load sections.
+
+  // This will validate (and possibly print) all data.
+  // TODO-LDML: we will be replacing this validation with validation-as-we-go below.
+  KMXPLUS_ASSERT(true, kmx::validate_kmxplus_data(kmxplusdata));
+
+  // Get out the SECT header
+  {
+    const KMX_DWORD offset = 0;
+    sect = kmx::as_kmxplus_sect(kmxplusdata+offset);
+    KMXPLUS_ASSERT(true, sect != nullptr);
+    // Specified data size fits in total
+    KMXPLUS_ASSERT(true, (offset+sect->header.size) < ex->kmxplus.dwKMXPlusSize);
+  }
 
-  if (validate_kmxplus_data(comp_keyboard)) {
-    _valid = true;
+  // Fill out the other sections we need.
+  {
+    const KMX_DWORD offset = sect->find(LDML_SECTIONID_STRS);
+    KMXPLUS_ASSERT(true, offset != 0);
+    KMXPLUS_ASSERT(true, offset < ex->kmxplus.dwKMXPlusSize);
+    strs = kmx::as_kmxplus_strs(kmxplusdata+offset);
+    KMXPLUS_ASSERT(true, strs != nullptr);
+    // Specified data size fits in total
+    KMXPLUS_ASSERT(true, (offset+strs->header.size) < ex->kmxplus.dwKMXPlusSize);
   }
+
+  {
+    const KMX_DWORD offset = sect->find(LDML_SECTIONID_KEYS);
+    KMXPLUS_ASSERT(true, offset != 0);
+    keys = kmx::as_kmxplus_keys(kmxplusdata+offset);
+    KMXPLUS_ASSERT(true, keys != nullptr);
+    // Specified data size fits in total
+    KMXPLUS_ASSERT(true, (offset+keys->header.size) < ex->kmxplus.dwKMXPlusSize);
+
+    // read all keys into array
+    for (KMX_DWORD i=0; i<keys->count; i++) {
+      const kmx::COMP_KMXPLUS_KEYS_ENTRY &entry = keys->entries[i];
+      KMX_DWORD len = 0;
+      KMX_WCHAR out[BUFSIZ];
+      if (entry.flags && LDML_KEYS_FLAGS_EXTEND) {
+        KMXPLUS_ASSERT(false, nullptr == strs->get(entry.to, out, BUFSIZ));
+        for(len=0; len<BUFSIZ && out[len]; len++);
+      } else {
+        UTF32 buf32[2];
+        buf32[0] = entry.to; // UTF-32
+        buf32[1] = 0; // to avoid UMR warning
+        UTF32 *sourceStart = &buf32[0];
+        const UTF32 *sourceEnd = &buf32[1]; // Reference off the end. NULL to avoid UMR.
+        UTF16 *targetStart = (UTF16*)out;
+        const UTF16 *targetEnd = (UTF16*)out+BUFSIZ-1;
+        ConversionResult result = ::ConvertUTF32toUTF16(&sourceStart, sourceEnd, &targetStart, targetEnd, strictConversion);
+        KMXPLUS_ASSERT(conversionOK, result);
+        *targetStart = 0;
```

Comment:
This conversion is a little painful. I wonder if we can use the unicode.h macros instead -- both faster and cleaner?

```
if(Uni_IsSMP(entry.to)) {
  out[0] = Uni_UTF32ToSurrogate1(entry.to);
  out[1] = Uni_UTF32ToSurrogate2(entry.to);
  out[2] = 0;
  len = 2;
} else {
  out[0] = (char16_t) entry.to;
  out[1] = 0;
  len = 1;
}
```

Reply:
It is painful. I have a separate ticket for improving this.

Reply:
https://github.com/keymanapp/keyman/issues/7134